### DISCUSSION TITLE
How detect that a certain block of work (a group of messages) were consumed, processed and acknowledged?

### DISCUSSION THREAD

**Discussion by hamidovz** [rabbitmq/rabbitmq-server#12204] [Other] [answered]
hi ! i am trying to build video processing app. what i want to do is , first process videos and convert them to 480p variant and if there is no more messages in current queue , then start to process 720p variants.

here is my current code

<details><summary>480p consumer</summary>

```
const amqp = require('amqplib');

async function process480p() {
  const conn = await amqp.connect('amqp://localhost');
  const channel = await conn.createChannel();
  
  const queue480p = '480p_queue';
  const queue720p = '720p_queue';
  
  await channel.assertQueue(queue480p, { durable: true });
  await channel.assertQueue(queue720p, { durable: true });
  await channel.prefetch(1,false)
  
  channel.consume(queue480p, async (msg) => {
    if (msg !== null) {

      const videoDetails = JSON.parse(msg.content.toString());
      console.log('Processing 480p for video:', videoDetails.videoId);

      // Simulate video processing
      await new Promise(resolve => setTimeout(resolve, 2000)); // 1 second processing time

      console.log('Processed 480p for video:', videoDetails.videoId);
      channel.ack(msg);

      // Check if the 480p queue is empty
      const { messageCount } = await channel.checkQueue(queue480p);
      if (messageCount === 0) {
        console.log('480p queue empty, enqueueing for 720p processing:', videoDetails.videoId);
        channel.sendToQueue(queue720p, Buffer.from(JSON.stringify(videoDetails)), { persistent: true });
      }
    }
  });
}

process480p();
```
</details>

<details><summary>720p consumer</summary>

```
const amqp = require('amqplib');

async function process720p() {
  const conn = await amqp.connect('amqp://localhost');
  const channel = await conn.createChannel();
  
  const queue720p = '720p_queue';
  await channel.assertQueue(queue720p, { durable: true });

  channel.consume(queue720p, (msg) => {
    if (msg !== null) {
      const videoDetails = JSON.parse(msg.content.toString());
      console.log('Processing 720p for video:', videoDetails.videoId);

      // Simulate video processing
      setTimeout(() => {
        console.log('Processed 720p for video:', videoDetails.videoId);
        channel.ack(msg);
      }, 1000); // 1 second processing time
    }
  });
}

process720p();
```
</details>

<details><summary>this is my publisher</summary>

```
const amqp = require('amqplib');

async function sendToQueue(videos) {
  const conn = await amqp.connect('amqp://localhost');
  const channel = await conn.createChannel();
  
  const queue480p = '480p_queue';
  await channel.assertQueue(queue480p, { durable: true });

  videos.forEach(videoDetails => {
    const message = JSON.stringify(videoDetails);
    channel.sendToQueue(queue480p, Buffer.from(message), { persistent: true });
    console.log('Sent video to 480p processing queue:', videoDetails);
  });

  await channel.close();
  await conn.close();
  
}

// Array of video details
const videoDetailsList = [
  { videoId: 101, path: '/path/to/video1.mp4' },
  { videoId: 102, path: '/path/to/video2.mp4' },
  { videoId: 103, path: '/path/to/video3.mp4' }
];

sendToQueue(videoDetailsList);
```
</details>

this is my results for 480p consumer from terminal:

```Processing 480p for video: 101
Processed 480p for video: 101
Processing 480p for video: 102
Processed 480p for video: 102
Processing 480p for video: 103
480p queue empty, enqueueing for 720p processing: 102
Processed 480p for video: 103
480p queue empty, enqueueing for 720p processing: 103
```

it looks like there is a problem about asynchronicity

**Comment by lukebakken** [comment]
> it looks like there is a problem about asynchronicity

No, it appears that everything is working as intended.

How do you define "empty" when a queue _might_ have more messages published to it?

When video 102 is done processing, the message for 103 must not have been in the queue yet. I don't know how `await channel.checkQueue(queue480p)` actually gets the current message count in the queue.

Your publisher should send a _DONE_ message when it is guaranteed that no more messages will be published.

**Reply by hamidovz** [reply]
@lukebakken  hi thank you for your answer ! but the thing is there is no guarantee that no more messages will be published. messages can be published every minute or every second and their publication time is uncertain. what can i do in this case ? or is it even possible to achieve what i want with rabbitmq ? if yes , how should i modify my code ? the only solution comes to my mind is using locking mechanism to make sure messages are consumed one after one synchronously

**Reply by lukebakken** [reply]
> but the thing is there is no guarantee that no more messages will be published

Then you can't really guarantee that the queue is ever empty, correct?

I think you should step back and focus on what you're trying to achieve. It sounds like you don't want to process 480p _and_ 720p videos at the same time. Why not something like this?

* Process 480p messages
* If, say, 10 seconds elapse since the last 480p message, switch to 720p mode.
* Process 720p messages.
* When done with 720p, return to 480p mode.

**Reply by hamidovz** [reply]
> > but the thing is there is no guarantee that no more messages will be published
> 
> Then you can't really guarantee that the queue is ever empty, correct?
> 
> I think you should step back and focus on what you're trying to achieve. It sounds like you don't want to process 480p _and_ 720p videos at the same time. Why not something like this?
> 
> * Process 480p messages
> * If, say, 10 seconds elapse since the last 480p message, switch to 720p mode.
> * Process 720p messages.
> * When done with 720p, return to 480p mode.

i understood. here is the full scenario what i am trying to do.  because of processing videos for 480p will be faster than 720p. i am trying to prioritize them. i want my app to start process videos for 480p and after checking current queue without any delay and finding out that current queue is empty ( even for 100ms ) then switch to 720p queue. But after consuming first message in 720p queue , check 480p queue again for new messages , if there is a new message , leave 720p queue and start to process 480p queue again. here is step by step process:

1. check 480p and if there are messages start consuming them
2. if there is no more messages for even 100ms ,  start consuming 720p variants of previous messages
3. when starting consuming for 720p variants , check 480p queue after every single message. if there is a message in 480p queue , leave 720p queue and get back to processing 480p queue.

my priority is 480p queue and if there is "free time" it should start process 720p variants but checking 480p everytime a single message consumed in 720p queue.

**Reply by lukebakken** [reply]
Why not do exactly what you state in 1, 2, 3?

* When 100ms passes without a message, cancel consuming from the 480p queue, and start consuming from the 720p queue
* After every 720p encoding, check the 480p message count
* If the 480p queue has messages, stop consuming from the 720p queue, and re-consume from the 480p queue

**Reply by hamidovz** [reply]
> Why not do exactly what you state in 1, 2, 3?
> 
> * When 100ms passes without a message, cancel consuming from the 480p queue, and start consuming from the 720p queue
> * After every 720p encoding, check the 480p message count
> * If the 480p queue has messages, stop consuming from the 720p queue, and re-consume from the 480p queue

i tried to do it but ended up race condition on message consuming , even i set up prefetch 1.  when i watched terminal , there result was something similar to this: 

```
Processing 480p for video: 101
Processed 480p for video: 101
Processing 480p for video: 102
480p queue empty, enqueueing for 720p processing
Processed 480p for video: 102
Processing 480p for video: 103
Processed 480p for video: 103
480p queue empty, enqueueing for 720p processing
```

**Reply by hamidovz** [reply]
> > Why not do exactly what you state in 1, 2, 3?
> > 
> > * When 100ms passes without a message, cancel consuming from the 480p queue, and start consuming from the 720p queue
> > * After every 720p encoding, check the 480p message count
> > * If the 480p queue has messages, stop consuming from the 720p queue, and re-consume from the 480p queue
> 
> i tried to do it but ended up race condition on message consuming , even i set up prefetch 1. when i watched terminal , there result was something similar to this:
> 
> ```
> Processing 480p for video: 101
> Processed 480p for video: 101
> Processing 480p for video: 102
> 480p queue empty, enqueueing for 720p processing
> Processed 480p for video: 102
> Processing 480p for video: 103
> Processed 480p for video: 103
> 480p queue empty, enqueueing for 720p processing
> ```

it seemed like second and first message attempt to check queue emptiness at the same time and i think it is because of asynchonous nature of javascript. i thought when i set prefetch 1  , no matter what happens messages will be consumed synchoronously.

**Reply by lukebakken** [reply]
> i thought when i set prefetch 1 , no matter what happens messages will be consumed synchoronously.

A prefetch of 1 means that RabbitMQ will only send your consumer 1 message, then wait for an ack, then send the next message, etc.

This is probably what is happening in your case:

* Your application receives a message, and processes it.
* The application sends the `ack` back to RabbitMQ, and then asks RabbitMQ for the count of "ready" messages in the queue. The count is 0 because the only other message is in the "Unacked" state (i.e., being processed currently)
* Your application then goes on to the 720p queue.

**Reply by hamidovz** [reply]
> Yes, this is always a possibility.

is there any repo or code for reference for situation similar to mine ? i can use external store but i just want to stick with rabbitmq itself. i am completely newbie to message queues

**Reply by lukebakken** [reply]
At this time, the best way for me to help you is for you to provide code I can run. Please provide a git repository I can clone with _detailed_ instructions for setting up your code and running it. Don't assume I know anything about node/javascript other than how to install node.js. Assume I have a running RabbitMQ server, but make everything else as easy as possible to get going. I don't have _any_ free time to fix syntax errors, that sort of thing.

I can fork your repository and poke around a bit.

**Reply by hamidovz** [reply]
> At this time, the best way for me to help you is for you to provide code I can run. Please provide a git repository I can clone with _detailed_ instructions for setting up your code and running it. Don't assume I know anything about node/javascript other than how to install node.js. Assume I have a running RabbitMQ server, but make everything else as easy as possible to get going. I don't have _any_ free time to fix syntax errors, that sort of thing.
> 
> I can fork your repository and poke around a bit.

thank you for your help and time ! here is the repo:  [rabbitmq sample app](https://github.com/hamidovz/rabbitmq-test)  i wrote instructions inside of the readme file. thank you again !

**Comment by michaelklishin** [comment]
This is a consumer coordination problem, how your consumers depend on each other or interact with each other is not something RabbitMQ is aware of or should be aware of.

What your applications should do when they are done with a certain stage is to publish a message of a special type to a separate queue. In general, in a pipeline of N different stages using N queues is guaranteed to be easier to reason about than one MegaQueue. Single Giant Shared Queue™ is likely the most problematic and the most infamous anti-pattern in messaging and the RabbitMQ community specifically.

Not that when a design problem does not fit the queue consumption-acknowledgement model, there's nothing wrong with [using a stream](https://www.rabbitmq.com/tutorials), for which there are now tutorials. Just an idea.

**Reply by hamidovz** [reply]
hI ! i ended up using basic.get and recursively check queue as there is no way to "pause" and "resume" queue consumption without hacky solutions.

**Reply by lukebakken** [reply]
Sure, that's probably not a bad decision in this case, even though we almost universally say "don't use `basic.get`".

**Reply by michaelklishin** [reply]
`basic.get` is a terrible idea. It defies all benefits of messaging. Absolutely to avoid besides basic tests.

Sounds like your system is designed with polling and ideas similar to polling in mind.

**Reply by hamidovz** [reply]
> `basic.get` is a terrible idea. It defies all benefits of messaging. Absolutely to avoid besides basic tests.
> 
> Sounds like your system is designed with polling and ideas similar to polling in mind.

so, what should i do instead ? because i have only 1 server and only 1 instance of ffmpeg can run at a time , and if there is message in 480p queue while the 720p queue is in process , 480p queue will start to consume and process it because there is no method available to pause a queue.  i have only 2 options here: if there is message in 480p while 720p queue in process , i can nack that message and requeue it and i should to this to all messages while 720p is in process. or second method is i should create a manager script and manage 2 queues from there but i need to somehow "stop" 480p queue while 720p queue in process and i think i should either use cancel method or close the connection of 480p completely.  and i need another queue to notify manager that 720p is done and make him open 480p queue again.  what else can i do ?

**Reply by hamidovz** [reply]
> > `basic.get` is a terrible idea. It defies all benefits of messaging. Absolutely to avoid besides basic tests.
> > Sounds like your system is designed with polling and ideas similar to polling in mind.
> 
> so, what should i do instead ? because i have only 1 server and only 1 instance of ffmpeg can run at a time , and if there is message in 480p queue while the 720p queue is in process , 480p queue will start to consume and process it because there is no method available to pause a queue. i have only 2 options here: if there is message in 480p while 720p queue in process , i can nack that message and requeue it and i should to this to all messages while 720p is in process. or second method is i should create a manager script and manage 2 queues from there but i need to somehow "stop" 480p queue while 720p queue in process and i think i should either use cancel method or close the connection of 480p completely. and i need another queue to notify manager that 720p is done and make him open 480p queue again. what else can i do ?

i thin i found another solution , while using normal quorum queue and consume method , i can create external locking mechanism ( for example using redis ) and requeue incoming messages of one queue until other queue releases lock. is it good idea to use locking with queues ? or is it anti pattern ?

**Reply by michaelklishin** [reply]
@hamidovz we will not design your apps and messaging topology for you. We do not have enough information to do so, you are not a paying customer or a regular contributor and above all, it is **your** job.

Sequentially processing N tasks when there's a shared resource like `ffmpeg` is a matter of creating N queues like

 * `video.encoding.inbox`
 * `video.encoding.480p`
 * `video.encoding.720p`
 * `video.encoding.1080p`
 
and adding a consumer (or a group of consumers with a [SAC](https://www.rabbitmq.com/docs/consumers#single-active-consumer)) to them.

Then the events are published to the `inbox`, which updates database records and publishes another message that's routed to `480p`, when that task is done, it publishes a message that's routed to `720p` and so on.

If you are certain you need to isolate `ffpmeg` access, use a key-value store of your choice or a "lock" of sorts in the data store you will use to keep track of task state, progress, and so on. It is a consumer coordination problem.

**Reply by michaelklishin** [reply]
@hamidovz using data services such as Consul, etcd or Redis (or anything else that you like) for consumer coordination is not an anti-pattern, it's exactly what you should do.

Limiting access to queues is not an anti-pattern either, sometimes it is a necessity of your workflow. That's why exclusive queues, Single Active Consumer and (in AMQP 0-9-1) exclusive consumers exist. I would not use exclusive consumers today when SAC exists and has been around for years, and exclusive queues are not meant to be (and cannot) be long lived but your thinking is largely in line with what I'm suggesting above.

A series of queues and consumers that publish (or republish) messages to kick off the next stage in the workflow, with SAC, a single channel with prefetch = 1, and an external data store helping you control access to shared resources, track progress, and so on.

This way there will be no shared queues, no polling, and a lot of freedom to configure individual consumers however you see fit (e.g. a 1080p consumer may use higher timeouts than the 480p one).

**Reply by lukebakken** [reply]
Just my $0.02, @hamidovz's use case is probably one of the only ones where I would say, "sure, use `basic.get`". It accomplishes what he needs and there is (apparently) not a huge message workload, so, why not?

**Reply by hamidovz** [reply]
> @hamidovz using data services such as Consul, etcd or Redis (or anything else that you like) for consumer coordination is not an anti-pattern, it's exactly what you should do.
> 
> Limiting access to queues is not an anti-pattern either, sometimes it is a necessity of your workflow. That's why exclusive queues, Single Active Consumer and (in AMQP 0-9-1) exclusive consumers exist. I would not use exclusive consumers today when SAC exists and has been around for years, and exclusive queues are not meant to be (and cannot) be long lived but your thinking is largely in line with what I'm suggesting above.
> 
> A series of queues and consumers that publish (or republish) messages to kick off the next stage in the workflow, with SAC, a single channel with prefetch = 1, and an external data store helping you control access to shared resources, track progress, and so on.
> 
> This way there will be no shared queues, no polling, and a lot of freedom to configure individual consumers however you see fit (e.g. a 1080p consumer may use higher timeouts than the 480p one).

i used locking mechanism and used redis for this. thank you for your time

**Reply by hamidovz** [reply, marked as answer]
> Just my $0.02, @hamidovz's use case is probably one of the only ones where I would say, "sure, use `basic.get`". It accomplishes what he needs and there is (apparently) not a huge message workload, so, why not?

i used locking mechanism and used redis for this. thank you for your time. but i didn't delete basic.get variant just in case.

**Comment by michaelklishin** [comment]
Also note that there is a [old book that just as relevant today](https://www.enterpriseintegrationpatterns.com/) on messaging patterns.

**Reply by hamidovz** [reply]
thank you very much i will check it out

**Reply by hamidovz** [reply]
@michaelklishin  which book among them do you recommend to read first ?

**Reply by michaelklishin** [reply]
There aren't that many books on messaging that are not the description of basics. There's [another one](https://www.amazon.com/Ultimate-Microservices-RabbitMQ-Architecture-Integration/dp/8197256306) by a RabbitMQ user from 2024 but Enterprise Integration Patterns has gotten more relevant over time, not less.
